core/src/databases/remote_databases/snowflake.rs

```diff
@@ -229,6 +217,60 @@ impl RemoteDatabase for SnowflakeRemoteDatabase {
         Ok((all_rows, schema))
     }
 
+    async fn _get_tables_used_by_query(
+        &self,
+        session: &SnowflakeSession,
+        query: &str,
+    ) -> Result<Vec<String>> {
+        let explain_query = format!("EXPLAIN {}", query);
+        let used_tables = session
+            .query(explain_query.clone())
+            .await?
+            .iter()
+            .filter_map(|row| match row.get::<String>("objects") {
+                Ok(objects) => Some(objects),
+                _ => None,
+            })
+            .collect();
+
+        Ok(used_tables)
+    }
+}
+
+#[async_trait]
+impl RemoteDatabase for SnowflakeRemoteDatabase {
+    async fn get_tables_used_by_query(&self, query: &str) -> Result<Vec<String>> {
+        let session = self._get_session().await?;
+        self._get_tables_used_by_query(&session, query).await
+    }
+
+    async fn execute_query(
+        &self,
+        tables: &Vec<Table>,
+        query: &str,
+    ) -> Result<(Vec<QueryResult>, TableSchema), QueryDatabaseError> {
+        let session = self._get_session().await?;
+
+        // Ensure that query only uses tables that are allowed.
+        let used_tables = self._get_tables_used_by_query(&session, query).await?;
+        let allowed_tables: HashSet<&str> = tables
+            .iter()
+            .filter_map(|table| table.remote_database_table_id())
+            .collect();
+
+        if used_tables
+            .iter()
+            .any(|table| !allowed_tables.contains(table.as_str()))
+        {
+            Err(QueryDatabaseError::ExecutionError(
+                "Query uses tables not allowed by the query plan".to_string(),
+            ))?
+        }
+
+        self._execute_query(&session, query).await
+    }
```

Comment:
I was initially puzzled by the get_tables + _get_tables.

(I understand that you created `_get_tables_used_by_query` to be able to use the same session in `execute_query`, right ?)

Taking a step back : I imagine that we want to ensure that remote database implements a function to get the tables used and that they are allowed.

Right now, the fact that we requires the methods to be implemented does not ensure that the methods will be called as we want them to be. The list of methods to implements is mostly an hint / reminder.

What about something like that (pseudo rust code):

```rust
trait RemoteDatabase {
    // Provide a default implementation to ensure that our checks are done.
    pub fn execute_query<T>(self, allowed_tables:&Vec<Table>, query:&str, session:Option<T>)
   -> Result(Vec<QueryResult>, TableSchema), QueryDatabaseError) {
let used_tables = self._get_tables_used_by_query(query, session).await?;
        let allowed_tables_ids: HashSet<&str> = tables
            .iter()
            .filter_map(|table| table.remote_database_table_id())
            .collect();

        if used_tables
            .iter()
            .any(|table| !allowed_tables_ids.contains(table.as_str()))
        {
            Err(QueryDatabaseError::ExecutionError(
                "Query uses tables not allowed by the query plan".to_string(),
            ))?
        }

        self._execute_query(query, session).await
    }

    // not public, to be implemented for each remote db
    async fn _get_tables_used_by_query<T>(query, session:Option<T>) -> Option<Vec<Table>>;
    async fn _execute_query<T>(query, session:Option<T>)  -> Result(Vec<QueryResult>, TableSchema), QueryDatabaseError);

}
```

Reply:
>     // not public, to be implemented for each remote db

I agree with you, but AFAIK trait methods are always public. And I don't want to force "session" in the interface as it might not make sense for other Remote DBs. Maybe `execute_query` could be renamed `greenlight_and_execute_query` and there could be a comment explaining that the method needs to check the query is allowed given the tables ?

Reply:
> I agree with you, but AFAIK trait methods are always public

Ah I see, however, it's not that important the "_" is enough imho.

> And I don't want to force "session" in the interface as it might not make sense for other Remote DBs. 

Don't you think using `Option` make it clear that it's not mandatory (in the end, you are just passing it around).

If I had to choose between an implementation where ensuring that the checking logic is done by a comment because we want to avoid exposing session vs an implementation that is more solid for the checking logic and "waste" sessions (by not passing it around), I would pick the later.

Obviously, to me the third option that I proposed above is the best of both worlds.

Reply:
But it makes the interface a lot less nice to use IMO. For some databases you won't need these private methods. 
I think a better name would go a long way -- `authorize_and_execute_query` ? If you don't implement or don't dig to understand the `authorize` part, it's on you isn't it ?

Reply:
> For some databases you won't need these private methods.

Hum, except the get_session, regardless of my proposal, for all remote database you need the methods, right ?

> But it makes the interface a lot less nice to use IMO

Care to elaborate ? To me it's nicer that the PR version with proxy methods with similar names.

> I think a better name would go a long way -- authorize_and_execute_query ? If you don't implement or don't dig to understand the authorize part, it's on you isn't it ?

A better name is indeed better but my personal take is that the more your code can enforce important rules by itself, the better. If I can avoid relying on the next dev (or future me) to prevent a mistake, I prefer to do it.

I am opinionated on this but it's your code and I'll let you implement the way you think it's the best. 

> but AFAIK trait methods are always public

After some dust-ing, it looks like the trait method inherit from the visibilty of the trait itself, you can split and put the private methods in a non public trait.

Reply:
> Hum, except the get_session, regardless of my proposal, for all remote database you need the methods, right ?

No I think it's specific to snowflake (others may need it too, but it's implementation details). In any case it should not be in the interface IMO.

> To me it's nicer that the PR version with proxy methods with similar names.

Those are not in the interface, and they aren't in the trait implementation. They are on the SnowflakeRemoteDatabase struct's impl (so they are entirely private)

Reply:
I really don't want to have snowflake implementation details leak into the RemoteDatabase interface. At the end of the day, what we really want from a Remote DB is ability to get schema for a set of tables and ability to authorize+execute a query. How a remote DB implements this is implementation details and shouldn't be in the interface.

Reply:
(it's of course my personal opinion on this, but strongly held)

Reply:
Will merge as-is for now (I renamed the method) -- we can always come back to it and change if we find a better solution

Reply:
(I have a second branch that's based on this one)

Reply:
> Hum, except the get_session, regardless of my proposal, for all remote database you need the methods, right ?

> No I think it's specific to snowflake (others may need it too, but it's implementation details). In any case it should not be in the interface IMO.

> Those are not in the interface, and they aren't in the trait implementation. They are on the SnowflakeRemoteDatabase struct's impl (so they are entirely private)

Ah ok, my bad I didn't read it correctly.

I'm pretty sure that all dbs would require theses 2 methods : a way to get the tables from a query and a way to execute a query which makes then sense to me to put "higher up" in the definition of RemoteDatabase, especially since we absolutely want make sure that the "this query will only use allowed tables" is not specific to Snowflake.

Reply:
> I don't want to have snowflake implementation details leak into the RemoteDatabase interface. At the end of the day, what we really want from a Remote DB is ability to get schema for a set of tables and ability to authorize+execute a query. How a remote DB implements this is implementation details and shouldn't be in the interface.

I see your point of view, I believe the nuance between you and me is that you leave the definition of "authorize" to the implementation where I see it as a non-implementation specific.

If we follow your pov, don't you think `get_tables_used_by_query` should not be part of RemoteDatabase ?

Reply:
>I f we follow your pov, don't you think get_tables_used_by_query should not be part of RemoteDatabase ?

This is a good point and I considered removing it...

At first, I wanted to only expose "authorize" and "execute" and have a global helper handle the `.authorize() + .execute()` but had to backtrack as I needed to keep a single session.

Maybe I could change  `get_tables_used_by_query`, call it `authorize_query` and have it take a list of tables ? that way it's obvious that `authorize_and_execute_query` needs to also do `authorize_query` ?

Reply:
> At first, I wanted to only expose "authorize" and "execute" and have a global helper handle the .authorize() + .execute() but had to backtrack as I needed to keep a single session.

So, it's more or less what I suggested, no ? (except that I added the optional session support)

> Maybe I could change get_tables_used_by_query, call it authorize_query and have it take a list of tables ? that way it's obvious that authorize_and_execute_query needs to also do authorize_query ?

Wouldn't that be half-way between the 2 "visions" ?
You'll still have a somewhat confusing Snowflake implementation where you rewrite the `authorize_query` to have a version with session etc..

I think it's better that we pick a way and be consistent, it makes the code more "obvious" in it's intent.

Reply:
I think it's fine to have private methods in the  implementation of your struct no ? The part that needs to be kept tidy is the trait.

I want to avoid Session in the implementation, the person using a remote database does not care about sessions, it cares about "get schema" and "execute query"